### PR TITLE
[DO NOT MERGE] remove hard coded openshift.local references from code

### PR DESCRIPTION
**Pull request by pweil-**
Proposal to remove the openshift.local* references in the cli command code.  The defaulter will allow @miminar to use a different base name when the changes for atomic/openshift based on binary are in.

@smarterclayton, @ashcrow - what do you think, necessary or overkill?

### CODE REVIEW

File: pkg/cmd/server/admin/create_bootstrappolicy_file.go [hunk 1]
```diff
@@ -16,12 +16,14 @@ import (
 
 	"github.com/openshift/origin/pkg/api/latest"
 	"github.com/openshift/origin/pkg/cmd/server/bootstrappolicy"
+	"github.com/openshift/origin/pkg/config/defaults"
```
Comment (smarterclayton) [inline pkg/cmd/server/admin/create_bootstrappolicy_file.go]: This is somewhat questionable - while DefaultPolicyFile was wierd before, this should really just be policy.json.  Don't try to overthink it.

File: pkg/cmd/server/admin/create_client.go [hunk 2]
```diff
@@ -69,7 +71,7 @@ func NewCommandCreateClient(commandName string, fullName string, out io.Writer)
 
 	flags.StringVar(&options.APIServerURL, "master", "https://localhost:8443", "The API server's URL.")
 	flags.StringVar(&options.PublicAPIServerURL, "public-master", "", "The API public facing server's URL (if applicable).")
-	flags.StringVar(&options.APIServerCAFile, "certificate-authority", "openshift.local.config/master/ca.crt", "Path to the API server's CA file.")
+	flags.StringVar(&options.APIServerCAFile, "certificate-authority", path.Join(defaults.NewDefaults().MasterConfigDir(), "ca.crt"), "Path to the API server's CA file.")
```
Comment (smarterclayton) [inline pkg/cmd/server/admin/create_client.go]: Too complicated - again this is something that I would argue we would just create in the current directory, or take a directory option.

File: pkg/cmd/server/admin/create_nodeconfig.go [hunk 3]
```diff
@@ -110,13 +111,15 @@ func NewCommandNodeConfig(commandName string, fullName string, out io.Writer) *c
 }
 
 func NewDefaultCreateNodeConfigOptions() *CreateNodeConfigOptions {
+	defaulter := defaults.NewDefaults()
+
 	options := &CreateNodeConfigOptions{SignerCertOptions: &SignerCertOptions{}}
-	options.VolumeDir = "openshift.local.volumes"
```
Comment (smarterclayton) [inline pkg/cmd/server/admin/create_nodeconfig.go]: This should be coming from the config defaulting mechanism, so it's not technically wrong.  However, the defaulter should be returning a server API object, not some new mechanism.

File: pkg/config/defaults/defaults.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+const (
+	// OpenShiftBaseName is the default base name for the OpenShift product.
+	OpenShiftBaseName = "openshift.local"
+	// AtomicBaseName is the default base name for the Atomic product.
+	AtomicBaseName = "atomic.local"
+)
+
+// Defaults provides utility for creating common directory names based on the base path
+// which may be a system default or an overrridden value.  See NewNewDefaultsFromBase.
+type Defaults struct {
```
Comment (smarterclayton) [inline pkg/config/defaults/defaults.go]: Why do we even need to distinguish between atomic and openshift?  Just use the "origin" directory as the default and have the installer lay down those.  We don't read from a directory by default so the installer laying down /etc/atomic and then pointing to it is fine.

Comment (ashcrow) [inline pkg/config/defaults/defaults.go]: :+1: to using "origin" instead of openshift or atomic.

**Comment by pweil-** [comment]
Ok, simpler version - just changed openshift.local to origin.local.  

Two things of note

1.  in `etc/server.go` the cluster name is changed - this didn't look like it would have side effects but I wanted to call it out just in case
1.  I left the old references to openshift.local in the `test/old-start-configs` alone

Will sweep through openshift-docs if this is good.

[test]

**Comment by smarterclayton** [comment]
> in etc/server.go the cluster name is changed - this didn't look like it would have side effects but I wanted to call it out just in case

It will break anyone who has old etcd data and doesn't also specify a cluster name.

**Comment by pweil-** [comment]
> It will break anyone who has old etcd data and doesn't also specify a cluster name.

revert or keep and document along the same lines as setting flags for using existing config?

**Comment by smarterclayton** [comment]
You can't provide a custom cluster name today.  I would say leave cluster name as is and add a TODO to it.

**Comment by pweil-** [comment]
reverted cluster name and rebased

re[test]

**Comment by liggitt** [comment]
This has a ton of backward compatibility implications.

**Comment by smarterclayton** [comment]
What implications does this have for the ansible installer?

**Comment by pweil-** [comment]
> This has a ton of backward compatibility implications.

Other than config file locations what issues do you see?  It was noted that most folks using the product already in a "real" sense would be pointing the start commands to specific directories already.

/cc @jwhonce for ansible implications

**Comment by pweil-** [comment]
@jwhonce - to my untrained eye it looks like ansible is already using `/etc/openshift`

**Comment by jwhonce** [comment]
@detiber or @sdodson have either of you seen this PR and/or started an update to openshift-ansible.

@twiest FYI

**Comment by sdodson** [comment]
I'll give it a test today, but I think we're immune to any problems created
by this change as we're explicitly setting all of those paths. However, we
may want to better understand where this is going in the future and see if
we need to adjust.

**Comment by liggitt** [comment]
I'm glad ansible and ansible-installed setups aren't affected, but do we not care about backwards compatibility with direct `openshift start` invocations?

**Comment by smarterclayton** [comment]
If we don't change it now we never will.  The alternative is to search on
disk for a set of directories.

On Jul 30, 2015, at 10:14 AM, Jordan Liggitt <notifications@github.com>
wrote:

I'm glad ansible and ansible-installed setups aren't affected, but do we
not care about backwards compatibility with direct openshift start
invocations?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3930#issuecomment-126343078>.

**Comment by pweil-** [comment]
I'd prefer to just have a release note that tells folks to rename the directory if they need to rather than add complexity to the code for something we don't intend on supporting going forward.  If others disagree I can make the change.

**Comment by ashcrow** [comment]
@pweil- I agree. As @smarterclayton said if we don't change it now we never will.

**Comment by smarterclayton** [comment]
I'm fine with the rename comment - we'll line up 1.0.5 around merging this

On Jul 31, 2015, at 11:13 AM, Paul Weil <notifications@github.com> wrote:

I'd prefer to just have a release note that tells folks to rename the
directory if they need to rather than add complexity to the code for
something we don't intend on supporting going forward. If others disagree I
can make the change.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3930#issuecomment-126721922>.

**Comment by smarterclayton** [comment]
Although let's have a brief discussion on whether "local.volumes" is better
than "origin.local.volumes"

On Jul 31, 2015, at 11:13 AM, Paul Weil <notifications@github.com> wrote:

I'd prefer to just have a release note that tells folks to rename the
directory if they need to rather than add complexity to the code for
something we don't intend on supporting going forward. If others disagree I
can make the change.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3930#issuecomment-126721922>.

**Comment by liggitt** [comment]
All three directories: config, volumes, etcd

**Comment by smarterclayton** [comment]
How about "cluster-config", "pod-volumes", and "cluster-etcd"

On Jul 31, 2015, at 11:24 AM, Jordan Liggitt <notifications@github.com>
wrote:

All three directories: config, volumes, etcd

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3930#issuecomment-126724196>.

**Comment by pweil-** [comment]
> Although let's have a brief discussion on whether "local.volumes" is better
> than "origin.local.volumes"

I don't find it better or worse since this is just a default location.  I would say that I prefer a standard prefix to them in terms of developing since I often blast the directories and it's easy with `rm something.*`  I assume that that is not uncommon.

**Comment by smarterclayton** [comment]
The defaults you set send a message about the alignment of the various
pieces - "origin" isn't something we make special necessarily.  You rarely
should be blowing away volumes since the kubelet manages it and will
cleanup on start.  You can delete cluster-* for your use case.

@liggitt, preferences?

On Fri, Jul 31, 2015 at 1:38 PM, Paul Weil <notifications@github.com> wrote:

> Although let's have a brief discussion on whether "local.volumes" is better
> than "origin.local.volumes"
>
> I don't find it better or worse since this is just a default location. I
> would say that I prefer a standard prefix to them in terms of developing
> since I often blast the directories and it's easy with rm something.* I
> assume that that is not uncommon.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3930#issuecomment-126765818>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by pweil-** [comment]
who do I need to pay off to say they like origin.local so I don't have to go back through these files? :moneybag: 

Just kidding, let's do what makes the best sense since we'll be living with it for ETERNITY.

**Comment by miminar** [comment]
:+1: `origin.local.*` defaults. It may happen that admin runs `openshift` by accident in workspace directory inappropriate for that purpose. `origin.local.*` are easy to spot and (re)move and they aren't likely to conflict with anything. Let the rpm spec and ansible choose better paths for `/etc/origin/` case.

**Comment by smarterclayton** [comment]
Ok, let's leave those as is.

On Aug 3, 2015, at 9:28 AM, Michal Minar <notifications@github.com> wrote:

[image: :+1:] origin.local.* defaults. It may happen that admin runs
openshift by accident in workspace directory inappropriate for that
purpose. origin.local.* are easy to spot and (re)move and they aren't
likely to conflict with anything. Let the rpm spec and ansible choose
better paths for /etc/origin/ case.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3930#issuecomment-127233423>.

**Comment by smarterclayton** [comment]
can you add the upgrade steps to UPGRADE.md about 1.0.5 and renaming your directories (what a user has to do to upgrade).

**Comment by pweil-** [comment]
rebased and added a 1.0.5 section to upgrade.md.  Ready for more review

**Comment by smarterclayton** [comment]
Ok, once I merge this can you send the email to dev@ informing people that this change has landed in master and will be in 1.0.5?  Probably forward that then to SME and other lists, it's likely to break quite a few people.

**Comment by danmcp** [comment]
@pweil- Needs a rebase

**Comment by pweil-** [comment]
closing this based on external discussions.  We will not be changing the `openshift.local` references for now.

**Pull request closed without merging**